### PR TITLE
Fix/join not found columns

### LINKED ISSUE #7: Cannot find columns using regular join
```scala
val df1 = spark.sql("SELECT * FROM (" +
      " SELECT COUNT(commit_hash) AS num, commit_hash" +
      " FROM ref_commits r" +
      " NATURAL JOIN commits c" +
      " GROUP BY commit_hash" +
      ") t WHERE num > 1")

val df2 = spark.sql("SELECT * FROM (" +
      " SELECT COUNT(c.commit_hash) AS num, c.commit_hash" +
      " FROM ref_commits r" +
      " INNER JOIN commits c" +
      " ON c.commit_hash = r.commit_hash" +
      " AND c.repository_id = r.repository_id" +
      " GROUP BY c.commit_hash" +
      ") t WHERE num > 1")
```

These two queries are the same, one with natural joins and the other with regular joins. Spark fails with the second one because it can't find some columns.

### PR DESCRIPTION
**Pull request by erizocosmico**
Depends on #11 which depends on #10 

Fixes #7 

Tests should be passing now. I can't ensure for other queries it will not fail. This was the result of fighting a lot against the spark built-in rules and the only way I could make all these queries work.

### CODE REVIEW

**Comment by erizocosmico** [comment]
Done. Because of the dedup, the columns were not correct. I've added a Project on top of each joined table to ensure it's still correct after deduping. Also added some more tests I forgot to add for some expressions.

**Comment by ajnavarro** [comment]
could you check Travis @erizocosmico ?

**Comment by erizocosmico** [comment]
It was a scalacheck error. Should be ok now

**Comment by ajnavarro** [comment]
@erizocosmico could you rebase please?

**Comment by erizocosmico** [comment]
Rebased

File: src/main/scala/tech/sourced/gitbase/spark/rule/PushdownJoins.scala [hunk 1]
```diff
@@ -64,6 +124,22 @@ case class JoinData(source: Option[DataSource] = None,
   */
 private[rule] object JoinOptimizer extends Logging {
 
+  private[rule] def hasSource(attr: Attribute): Boolean = {
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/gitbase/spark/rule/PushdownJoins.scala]: You can remove { and } here

File: src/main/scala/tech/sourced/gitbase/spark/rule/PushdownJoins.scala [hunk 2]
```diff
@@ -64,6 +124,22 @@ case class JoinData(source: Option[DataSource] = None,
...
+    getSource(attr) != ""
+  }
+
+  private[rule] def getSource(attr: NamedExpression): String = {
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/gitbase/spark/rule/PushdownJoins.scala]: also here

**Comment by erizocosmico** [comment]
Done @ajnavarro

**Pull request merged**